var/spack/repos/builtin/packages/arrow/package.py

```diff
@@ -63,37 +96,49 @@ def patch(self):
-            args.append("-DARROW_CUDA:BOOL=OFF")
+        filter_file(
+            r'set\(ARROW_LLVM_VERSIONS "10" "9" "8" "7"\)',
+            'set(ARROW_LLVM_VERSIONS "11" "10" "9" "8" "7")',
+            "cpp/CMakeLists.txt",
+            when="@:2.0.0",
+        )
 
-        if self.spec.satisfies("+python"):
-            args.append("-DARROW_PYTHON:BOOL=ON")
-        else:
-            args.append("-DARROW_PYTHON:BOOL=OFF")
+        filter_file(
+            r"#include <llvm/Support/DynamicLibrary\.h>",
+            r"#include <llvm/Support/DynamicLibrary.h>" + "\n" + r"#include <llvm/Support/Host.h>",
+            "cpp/src/gandiva/engine.cc",
+            when="@2.0.0",
+        )
 
-        if self.spec.satisfies("+parquet"):
-            args.append("-DARROW_PARQUET:BOOL=ON")
-        else:
-            args.append("-DARROW_PARQUET:BOOL=OFF")
+    def cmake_args(self):
+        args = ["-DARROW_DEPENDENCY_SOURCE=SYSTEM", "-DARROW_NO_DEPRECATED_API=ON"]
 
-        if self.spec.satisfies("+orc"):
-            args.append("-DARROW_ORC:BOOL=ON")
+        if self.spec.satisfies("+shared"):
+            args.append(self.define("BUILD_SHARED", "ON"))
         else:
-            args.append("-DARROW_ORC:BOOL=OFF")
+            args.append(self.define("BUILD_SHARED", "OFF"))
+            args.append(self.define("BUILD_STATIC", "ON"))
+
+        if self.spec.satisfies("@:0.11.99"):
+            # ARROW_USE_SSE was removed in 0.12
+            # see https://issues.apache.org/jira/browse/ARROW-3844
+            args.append(self.define("ARROW_USE_SSE", "ON"))
+
+        args.append(self.define_from_variant("ARROW_COMPUTE", "compute"))
+        args.append(self.define_from_variant("ARROW_CUDA", "cuda"))
+        args.append(self.define_from_variant("ARROW_GANDIVA", "gandiva"))
+        args.append(self.define_from_variant("ARROW_GLOG", "glog"))
+        args.append(self.define_from_variant("ARROW_HDFS", "hdfs"))
+        args.append(self.define_from_variant("ARROW_IPC", "ipc"))
+        args.append(self.define_from_variant("ARROW_JEMALLOC", "jemalloc"))
+        args.append(self.define_from_variant("ARROW_ORC", "orc"))
+        args.append(self.define_from_variant("ARROW_PARQUET", "parquet"))
+        args.append(self.define_from_variant("ARROW_PYTHON", "python"))
+        args.append(self.define_from_variant("ARROW_TENSORFLOW", "tensorflow"))
+        args.append(self.define_from_variant("ARROW_WITH_BROTLI", "brotli"))
+        args.append(self.define_from_variant("ARROW_WITH_LZ4", "lz4"))
+        args.append(self.define_from_variant("ARROW_WITH_SNAPPY", "snappy"))
+        args.append(self.define_from_variant("ARROW_WITH_ZLIB", "zlib"))
+        args.append(self.define_from_variant("ARROW_WITH_ZSTD", "zstd"))
 
-        for dep in ("flatbuffers", "rapidjson", "snappy", "zlib", "zstd"):
-            args.append("-D{0}_HOME={1}".format(dep.upper(), self.spec[dep].prefix))
-        args.append("-DZLIB_LIBRARIES={0}".format(self.spec["zlib"].libs))
```

Comment:
Are these path arguments no longer needed?

Reply:
in the stack I'm using arrow in, it works without, but indeed in that stack I don't use e.g. snappy or zstd. 
I will check...

Reply:
ok, `spack install arrow +zstd+zlib+snappy` works but `spack install arrow@8 +zstd+zlib+snappy` complains about not finding Snappy indeed... 
Will put those lines back.

Reply:
well, actually I'm not so sure `snappy` was working at all in `arrow@8` with the previous recipe. Seems snappy was a declared as dependency but in arrow's CMakeLists it was not requested, unless explicitly added with `ARROW_WITH_SNAPPY` ...

Reply:
@tldahlgren instead of checking that the new variants I'm introducing works with previous versions, would it be acceptable if I only add them for "@9:" ?

And in the same spirit get the 

```
for dep in ("flatbuffers", "rapidjson", "snappy", "zlib", "zstd"):
            args.append("-D{0}_HOME={1}".format(dep.upper(), self.spec[dep].prefix))
        args.append("-DZLIB_LIBRARIES={0}".format(self.spec["zlib"].libs))
```

lines protected by a `with when("@:8"):` ?

Reply:
Yes, you definitely want to add version constraints to ensure constraints are only associated with the package when they are known to apply.

Reply:
ok, I've added back the `args.append(...` lines for version < 9. 
I've kept the addition of the variants untouched though, as I believe their default values are "as before".